src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -119,25 +207,147 @@ bool DeepChangeChecker::check_row(Table const& table, ObjKeyType key, size_t dep
+        for (const auto& key_path : key_path_array) {
+            auto next_object_key_to_check = object_key;
+            for (size_t i = 0; i < key_path.size(); i++) {
+                // Check for a change on the current depth.
+                auto table_key = key_path[i].first;
+                auto column_key = key_path[i].second;
+                auto iterator = m_info.tables.find(table_key.value);
+                if (iterator != m_info.tables.end() &&
+                    iterator->second.modifications_contains(next_object_key_to_check, {column_key})) {
+                    return true;
+                }
+
+                // Advance one level deeper into the key path if possible.
+                if (i < key_path.size() - 1) {
+                    auto column_type = column_key.get_type();
+                    if (column_type == col_type_Link) {
+                        const Obj obj = m_root_table.get_object(ObjKey(next_object_key_to_check));
+                        next_object_key_to_check = obj.get<ObjKey>(ColKey(column_key)).value;
+                    }
+                }
+            }
+        }
+    }
+
+    return false;
+}
+
+ObjectChangeChecker::ObjectChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                         std::vector<RelatedTable> const& related_tables,
+                                         const std::vector<KeyPathArray>& key_path_arrays)
+    : DeepChangeChecker(info, root_table, related_tables, key_path_arrays)
+{
+}
+
+std::vector<int64_t> ObjectChangeChecker::operator()(ObjKeyType object_key)
+{
+    std::vector<int64_t> changed_columns = {};
+
+    for (const auto& key_path_array : m_key_path_arrays) {
+        for (const auto& key_path : key_path_array) {
+            check_key_path(changed_columns, key_path, 0, m_root_table, object_key);
+        }
+    }
+
+    return changed_columns;
+}
+
+void ObjectChangeChecker::check_key_path(std::vector<int64_t>& changed_columns, const KeyPath& key_path, size_t depth,
+                                         const Table& table, const ObjKeyType& object_key_value)
+{
+    if (depth >= key_path.size()) {
+        // We've reached the end of the key path.
+        return;
+    }
+
+    auto table_key = key_path[depth].first;
+    auto column_key = key_path[depth].second;
+
+    // Check for a change on the current depth level.
+    auto iterator = m_info.tables.find(table_key.value);
+    if (iterator != m_info.tables.end() && iterator->second.modifications_contains(object_key_value, {column_key})) {
+        // If an object linked to the root object was changed we only mark the
+        // property of the root objects as changed.
+        auto root_column_key = key_path[0].second;
+        changed_columns.push_back(root_column_key.value);
+    }
+
+    // Advance one level deeper into the key path.
+    auto column_type = column_key.get_type();
+    if (column_type == col_type_Link) {
```

Comment:
I think link lists should be handled as well.

Reply:
Fixed with f923a25.